impl/src/main/java/com/okta/sdk/impl/client/DefaultClientBuilder.java

```diff
@@ -319,6 +359,53 @@ else if (this.clientConfig.getClientCredentialsResolver() == null) {
             this.clientConfig.setBaseUrlResolver(new DefaultBaseUrlResolver(this.clientConfig.getBaseUrl()));
         }
 
+        // OAuth2
+        if (this.clientConfig.getAuthenticationScheme() == AuthenticationScheme.OAUTH2) {
+            if (this.clientConfig.getAuthorizationMode() != null) {
+                if (this.clientConfig.getAuthorizationMode().equals("PrivateKey")) {
+                    // Validate client config
+                    validateOAuth2ClientConfig(this.clientConfig);
+
+                    // Get access token asynchronously
+                    CompletableFuture<String> completableFuture = CompletableFuture.supplyAsync(() -> {
```

Comment:
This should be pulled out of the ClientBuilder, and likely put in one of the new OAuth specific classes.
I'm not sure if we should even prefetch the token, but that is a different discussion ;)

Reply:
It looks like only one token is ever retrieved? once this token expires where is the logic to get a new one?

This logic will need some good tests.

There are a lot of lost exceptions in here too that worry me, when the `ClientCredentials` does not have a valid token that `ClientCredentials` implementation should throw an exception.

It's also possible this is a recoverable event, i.e. drop in network connection or something.

My suggestion is to get this logic working on-demand as needed (as in when the tokens expire), and then optimize (pre-fetch, and async update). As the optimize use cases would still likely need the basic implementation.

We also need to be careful when using the default thread pools as they may use a different classloader on some systems (i.e. OSGi).

Reply:
Noted, see updated code. The updated code will retrieve access token for every API call. This will need to be optimized later to fetch only on expiry and reusing valid tokens until they expire.

Reply:
Working on the token management code.. will push shortly.

Reply:
Implemented access token refresh on expiry. Please take a look.

Reply:
Pushed with IT & UTs.